site/src/xServices/workspaces/workspacesXService.ts

```diff
@@ -277,48 +298,38 @@ export const workspacesMachine = createMachine(
 
         workspaceRef.send("UPDATE_VERSION")
       },
-      // Opened discussion on XState https://github.com/statelyai/xstate/discussions/3406
-      updateWorkspaceRefs: assign({
-        workspaceRefs: (context, event) => {
-          let workspaceRefs = context.workspaceRefs
-
-          if (!workspaceRefs) {
-            throw new Error("No workspaces loaded.")
-          }
-
-          // Update the existent workspaces or create the new ones
-          for (const data of event.data) {
-            const ref = workspaceRefs.find((ref) => ref.id === data.id)
-
-            if (!ref) {
-              workspaceRefs.push(spawn(workspaceItemMachine.withContext({ data }), data.id))
-            } else {
-              ref.send({ type: "UPDATE_DATA", data })
-            }
-          }
-
-          // Remove workspaces that were deleted
-          for (const ref of workspaceRefs) {
-            const refData = event.data.find((workspaceData) => workspaceData.id === ref.id)
-
-            // If there is no refData, it is because the workspace was deleted
-            if (!refData) {
-              // Stop the actor before remove it from the array
-              if (ref.stop) {
-                ref.stop()
-              }
-
-              // Remove ref from the array
-              workspaceRefs = workspaceRefs.filter((oldRef) => oldRef.id !== ref.id)
-            }
-          }
-
-          return workspaceRefs
+      assignUpdatedWorkspaceRefs: assign({
+        workspaceRefs: (_, event) => {
+          const newWorkspaceRefs = event.data.newWorkspaces.map((workspace) =>
+            spawn(workspaceItemMachine.withContext({ data: workspace }), workspace.id),
+          )
+          return event.data.refsToKeep.concat(newWorkspaceRefs)
         },
       }),
     },
     services: {
       getWorkspaces: (context) => API.getWorkspaces(queryToFilter(context.filter)),
+      updateWorkspaceRefs: (context, event) => {
```

Comment:
I think I can understand what is happening here but I would add a comment about the `refsToKeep` and what the `ref.stop` is doing. Also, do we need to make this a service since it does not run any async ops.

Reply:
Yeah, I made it a service because we don't want to do `stop` and `send` in an `assign` action, and it's okay in another action, but then it's harder to send data from one action to the next (which is by design, I think). I think I would need to save `refsToKeep` and `newWorkspaces` in context, and that felt like clutter.